File: Programmers/Python3/Level 3/야근 지수/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/12927

from bisect import bisect_left


def solution(n, works):
    total = sum(works)
    if total <= n:  # 야근전에 모든 일을 다 마칠수있는경우
        return 0

    works.sort()
    # 최대값을 최대한 줄임
    while n:
        max_num = works[-1]
        left = bisect_left(works, max_num)
        for idx in range(left, len(works)):
            n -= 1
            works[idx] -= 1
            if n == 0:
                break

    score = 0
    for num in works:
        score += num ** 2

    return score

File: Programmers/Javascript/Level 3/야근 지수/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/12927?language=javascript

function solution(n, works) {
  const totalWorkTime = works.reduce((prev, cur) => prev + cur, 0);
  let answer = 0;

  if (totalWorkTime <= n) {
    return answer;
  }

  works.sort((a, b) => b - a);

  // 가장 오래걸리는 작업부터 줄여나감
  while (n > 0) {
    const maxWork = works[0];

    for (let idx = 0; idx < works.length; idx++) {
      if (works[idx] !== maxWork || n <= 0) {
        break;
      }

      works[idx] -= 1;
      n -= 1;
    }
  }

  return works.reduce((prev, cur) => prev + cur ** 2, 0);
}
